### ISSUE TITLE
[ARM64] Test failed: Loader/classloader/TypeGeneratorTests/TypeGeneratorTest1183/Generated1183/Generated1183.sh

### ISSUE THREAD

**Issue by VincentBu** [dotnet/coreclr#24263]
**Job:**
https://ci.dot.net/job/dotnet_coreclr/job/master/view/x64/job/jitstress/job/x64_checked_centos7.1_r2r_gcstress15_tst/138/

**Failed tests:**
Loader/classloader/TypeGeneratorTests/TypeGeneratorTest1183/Generated1183/Generated1183.sh

**Log:**
```
Loader/classloader/TypeGeneratorTests/TypeGeneratorTest1183/Generated1183/Generated1183.sh [FAIL]
Assert failure(PID 125525 [0x0001ea55], Thread: 125525 [0x1ea55]): SanityCheck()
```

**Comment by VincentBu** [comment]
The below test failed with same assertion:
https://mc.dot.net/#/user/coreclr-outerloop-gcstress-extra/ci~2Fdotnet~2Fcoreclr~2Frefs~2Fheads~2Fmaster/test~2Ffunctional~2Fcli~2F/20190608.1/workItem/JIT.Regression/analysis/xunit/JIT_Regression._JitBlue_GitHub_23199_GitHub_23199_GitHub_23199_~2F_JitBlue_GitHub_23199_GitHub_23199_GitHub_23199_sh

**Comment by RussKeldorph** [comment]
These seems to be consistently hitting `GitHub_23199` for Linux/ARM64 `gcstress0xc_jitminopts_heapverify1`.

```
Assert failure(PID 4067 [0x00000fe3], Thread: 4067 [0x0fe3]): SanityCheck()
    File: /__w/4/s/src/vm/methodtable.cpp Line: 10110
    Image: /root/helix/work/correlation/corerun

/root/helix/work/workitem/JIT/Regression/JitBlue/GitHub_23199/GitHub_23199/GitHub_23199.sh: line 256:  4067 Aborted                 (core dumped) $LAUNCHER $ExePath \"${CLRTestExecutionArguments[@]}\"

Return code:      1
Raw output file:      /root/helix/work/workitem/JIT/Regression/Reports/JIT.Regression/JitBlue/GitHub_23199/GitHub_23199/GitHub_23199.output.txt
Raw output:
BEGIN EXECUTION
/root/helix/work/correlation/corerun GitHub_23199.exe ''
40
Expected: 100
Actual: 134
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:
> set CORE_ROOT=/root/helix/work/correlation
> /root/helix/work/workitem/JIT/Regression/JitBlue/GitHub_23199/GitHub_23199/GitHub_23199.sh
Expected: True
Actual:   False
```

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
Trying to repro the arm64 issue...

**Comment by AndyAyersMS** [comment]
Note GitHub_23199 also sets `COMPlus_GcStressOnDirectCalls` in its execution script. And once you do this, it fails repeatedly.

The top of thread issue in the TypeGeneratorTest is probably unrelated.

**Comment by AndyAyersMS** [comment]
This is crashing at the second call in `GetTwoPointersStruct`. Looks like in `MethodDesc::ParseReturnKindFromSig` the runtime is not classifying `Span<T>` as something that potentially requires return value protection on ARM64.. So the first call in the method (which returns `Span<byte>` in `x0` and `x1` where `x0` is an interior pointer) fails to GC protect `x0` and it becomes stale, and the app then fails when the stack slot `x0` gets stored into after the first call is reported live at that second call.

`MethodDesc::ParseReturnKindFromSig` needs to handle `IsByrefLike` as a special case, if we have multi reg return structs and are not handling the x64 ABI. However the "easy" fix is to return `RT_Illegal` and while it would prevent us from breaking in for GC stress for a case like this, I wonder if it would leave a potential GC hole...

cc @sandreenko

**Comment by sandreenko** [comment]
After recent changes we do not use `ParseReturnKindFromSig` for correctness (actually do, but only in one place `ComPlusMethodFrame::GcScanRoots` that is only for special known internal functions and it has asserts that it doesn't meet struct returns).

For correctness we now always use `GCInfo v2`  (`ReturnKind GetReturnKind(Thread *pThread, EECodeInfo *codeInfo)`), so this bug can't lead to GC halls.

> However the "easy" fix is to return RT_Illegal and while it would prevent us from breaking in for GC stress for a case like this.

That sounds good.

**Comment by AndyAyersMS** [comment]
I don't have a CentOS machine handy, so will first try and repro the top issue on ubuntu.

**Comment by xiangzhai** [comment]
Hi,

Base on [commit e999f37a83583768aa68ad9fda805088ac9df6b8](https://github.com/dotnet/coreclr/commit/e999f37a83583768aa68ad9fda805088ac9df6b8)  , without @AndyAyersMS  `s [patch](https://github.com/dotnet/coreclr/pull/25333)   , mips64el is also able to reproduce such issue:

```
export CORE_LIBRARIES=/home/loongson/zhaixiang/corefx-master-Linux.mips64.Debug
export COMPlus_LogEnable=1
export COMPlus_LogToConsole=1
export COMPlus_LogLevel=10
export COMPlus_JitDump=*
export COMPlus_JitFunctionTrace=1

./bin/Product/Linux.mips64.Debug/corerun HelloWorld.dll

...
TID 03a3: Looking up System.ByReference`1 by name.                              
TID 03a3: PHASEDLOAD: LoadTypeHandleForTypeKey for type System.ByReference`1 to level LOADED
TID 03a3: PHASEDLOAD: table contains:                                           
TID 03a3: LoadTypeHandle: Loading Class from Module 000000FF7C680020 token 20000b1
TID 03a3: ConstructCrst with this:0x2ba39950                                    
TID 03a3: PHASEDLOAD: Creating loading entry for type System.ByReference`1          
TID 03a3: PHASEDLOAD: About to do incremental load of type System.ByReference`1 (0000000000000000) from level BEGIN
TID 03a3: Loading class "System.ByReference`1" from module "/home/loongson/zhaixiang/gerrit/coreclr-mips64-dev/bin/Product/Linux.mips64.Debug/System.Private.CoreLib.dll" in domain 0x000000012BA01EB0 System Domain
TID 03a3: SD: MT::MethodIterator created for System.ValueType.                  
TID 03a3: EEC::IMD: pNewMD:0x7c695ce8 for tok:0x6000ef1 (System.ByReference`1[T]::.ctor)
TID 03a3: EEC::IMD: pNewMD:0x7c695d20 for tok:0x6000ef2 (System.ByReference`1[T]::get_Value)
TID 03a3: STATICS: Placing statics for System.ByReference`1[T]                  
TID 03a3: STATICS: Static field bytes needed (0 is normal for non dynamic case)0
TID 03a3: STATICS: Placing ThreadStatics for System.ByReference`1[T]            
TID 03a3: STATICS: ThreadStatic field bytes needed (0 is normal for non dynamic case)0
TID 03a3: Placing 0 statics (0 handles) for class System.ByReference`1[T].          

Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()

```

Then applied @AndyAyersMS 's [patch](https://github.com/dotnet/coreclr/pull/25333)  , no SEGFAULT any more,  but similar assertion failure as @RussKeldorph mentioned:

```
Assert failure(PID 32254 [0x00007dfe], Thread: 32254 [0x7dfe]): SanityCheck()   
    File: /home/loongson/zhaixiang/gerrit/coreclr-mips64-dev/src/vm/methodtable.cpp Line: 10110
    Image: /home/loongson/zhaixiang/gerrit/coreclr-mips64-dev/bin/Product/Linux.mips64.Debug/corerun

```

[Detailed mips64.debug.log](https://github.com/dotnet/coreclr/files/3328009/mips64.debug.log)

Thanks,
Leslie Zhai

**Comment by AndyAyersMS** [comment]
@xiangzhai the `SanityCheck` assert is fairly generic, it means some object that should be a MethodTable doesn't appear to be one. In your case we are very early in runtime startup so I would suspect some issue with class layout or the way the very early classes are constructed or statics are populated or similar? Suggest you open a new issue to track the status of bringup on Mips64.

I can't reproduce the Generated1183 errors, so I'm going to close this issue.

**Comment by xiangzhai** [comment]
Hi @AndyAyersMS 

Thanks for your kind reponse!

It is not reproducible when merged with upstream/master the [commit 6ab02e1cc487c64fb733fbb9d1be653337691aff](https://github.com/dotnet/coreclr/commit/6ab02e1cc487c64fb733fbb9d1be653337691aff) for mips64el.

Cheers,
Leslie Zhai
